### PR TITLE
Restrict permission check on repositories and fix some problems

### PR DESCRIPTION
**Pull request by lunny**
This PR will refactor permission check of repositories, especially units permission check. It will restrict permission check on organization repositories and fix some bugs. It creates a new struct `Permission` and a function to retrieve permission from `User` and `Repository`. 

You can know all the permission information from `Permission` and it has been saved on `context.Context` when middleware `repoAssignment` was called. After this PR,  we removed some functions like `IsRepositoryWriter` because we don't know that if the repo is on an organization and there are some teams give permission to the user. Instead it provides `CanAccess(unit)` and `CanWrite(unit)` to determine if the `User` could read or write to `code`, `issues`, `releases` or other units.

This also give a break change about a private repository on an organization. When a user on a team has write permission to a repository but he also is a collaborator with read right to that repository. Before this PR, the team settings will be ignored, but in this PR, the higher permission will be given to the user. Don't know it's a bug or a break change. I added the breaking label.

### CODE REVIEW

File: routers/api/v1/repo/collaborators.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ func ListCollaborators(ctx *context.APIContext) {
 	// responses:
 	//   "200":
 	//     "$ref": "#/responses/UserList"
-	if !ctx.Repo.IsWriter() {
```
Comment (techknowlogick) [inline routers/api/v1/repo/collaborators.go:37]: Gitea copyright

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -151,6 +151,7 @@ func repoAssignment() macaron.Handler {
 			return
 		}
 		repo.Owner = owner
+		ctx.Repo.Repository = repo
```
Comment (techknowlogick) [inline routers/api/v1/api.go:154]: Gitea copyright

File: routers/api/v1/repo/file.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func GetRawFile(ctx *context.APIContext) {
 	// responses:
 	//   200:
 	//     description: success
-	if !ctx.Repo.HasAccess() {
```
Comment (techknowlogick) [inline routers/api/v1/repo/file.go:41]: Gitea copyright

File: routers/api/v1/repo/issue_label.go [hunk 4]
```diff
@@ -90,11 +90,6 @@ func AddIssueLabels(ctx *context.APIContext, form api.IssueLabelsOption) {
 	// responses:
 	//   "200":
 	//     "$ref": "#/responses/LabelList"
-	if !ctx.Repo.IsWriter() {
```
Comment (techknowlogick) [inline routers/api/v1/repo/issue_label.go:93]: Gitea copyright

File: routers/api/v1/repo/label.go [hunk 5]
```diff
@@ -123,7 +123,7 @@ func CreateLabel(ctx *context.APIContext, form api.CreateLabelOption) {
 	// responses:
 	//   "201":
 	//     "$ref": "#/responses/Label"
-	if !ctx.Repo.IsWriter() {
+	if !ctx.Repo.CanWrite(models.UnitTypeIssues) && !ctx.Repo.CanWrite(models.UnitTypePullRequests) {
```
Comment (techknowlogick) [inline routers/api/v1/repo/label.go]: Gitea copyright

**Review by techknowlogick — CHANGES_REQUESTED** [review]
I didn't check all files, but Gitea copyright needs to be added in some files.

File: models/repo_permission.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+			}
+		}
+
+		if !found && !repo.IsPrivate {
+			perm.UnitsMode[u.Type] = AccessModeRead
+		}
+	}
+
+	perm.Units = make([]*RepoUnit, 0, len(repo.Units))
+	for t, _ := range perm.UnitsMode {
```
Comment (zeripath) [inline models/repo_permission.go]: Lint doesn't like the second argument

File: models/repo_permission.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	if err = repo.getOwner(e); err != nil {
+		return
+	}
+	if !repo.Owner.IsOrganization() {
+		return
+	}
+
+	// Collaborators on organization repos will not be limited by teams units
+	if isCollaborator, err := repo.isCollaborator(e, user.ID); err != nil {
+		return perm, err
```
Comment (zeripath) [inline models/repo_permission.go:141]: All of the other returns rely on the default named return values.

File: models/repo_permission.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	} else if isCollaborator {
+		return perm, nil
```
Comment (zeripath) [inline models/repo_permission.go]: Similarly.

File: routers/api/v1/api.go [hunk 9]
```diff
@@ -1,4 +1,5 @@
 // Copyright 2015 The Gogs Authors. All rights reserved.
+// Copyright 2016 The Gitea Authors. All rights reserved.
```
Comment (zeripath) [inline routers/api/v1/api.go:2]: 2018

**Review by zeripath — CHANGES_REQUESTED** [review]

File: routers/api/v1/api.go (hunk 9 above)
Comment (techknowlogick) [reply routers/api/v1/api.go:2]: I think maybe 2016 because the file might have been modified then.

Comment (zeripath) [reply routers/api/v1/api.go:2]: OK

Comment (lunny) [reply routers/api/v1/api.go:2]: @techknowlogick @zeripath yes. If I explicitly know the file is modified on some year I will change to that year, but will change it to 2018.

File: models/repo_permission.go (hunk 6 above)
Comment (lunny) [reply models/repo_permission.go]: done.

File: models/repo_permission.go (hunk 7 above)
Comment (lunny) [reply models/repo_permission.go:141]: Because this `err` is a local variable not that `err` return variable.

**Comment by kTitan** [comment]
Does this PR also fix #4329?

**Comment by lunny** [comment]
Just place a FIXME and another PR could change that function. This PR will only refactor or fix permissions check.

**Comment by lunny** [comment]
I think this is ready for review

**Comment by lunny** [comment]
rebased

**Review by techknowlogick — APPROVED** [review]

**Comment by lunny** [comment]
@zeripath need your approval.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
@lunny please resolve conflicts

**Comment by lunny** [comment]
@lafriks done.

**Review by zeripath — APPROVED** [review]

**Pull request merged**